### PR TITLE
RAC-4783: User EMC Internal Artifactory as the Debian Staging

### PR DESCRIPTION
**Pull request by panpan0000**
depends on https://github.com/RackHD/on-build-config/pull/258

reopen for https://github.com/RackHD/on-build-config/pull/156

**Background**
https://rackhd.atlassian.net/browse/RAC-4783

in pipeline, Between deb-build and ova/vagrant-build, there should be somewhere to store the temp debian packages.
aka, after deb-build done, the build packages should be in a deb repo registry,
so in next ova or vagrant build, they can apt-get install the packages from above deb repo registry.
image

Previously, bintray rackhd-mirror is used to serve this purpose. but it's not how bintray should be used.

**Solution**
use EMC Internal MN Lab Artifactory as the replacement.
say, 
(a) after deb-build done, upload packages to artifactory
(b) during packer build, use artifactory as the apt-source

**Review Tips**
soooo long PR to be review ?? No....

(1) Below files are **copied** from original OnRack EOS2GitHub repo (```OnRack/release-tools/lib```) 😁 
```
build-release-tools/lib/ArtifactoryTools.py
build-release-tools/lib/DebianToolsOnArtifactory.py
build-release-tools/lib/resultTool.py
```
add a is_version_exist() function in  ArtifactoryTools.py

(2) below files is modified based on ```OnRack/release-tools/bin/make-debian-release.py```
```build-release-tools/application/upload_staging_deb_to_artifactory.py```

(3) below two files are modified to use new script of above and new creds of MN artifactory
```
jobs/build_debian/build_debian.groovy
jobs/build_debian/build_debian.sh
```

(4) Modify OVA/vagrant/docker where  the staging apt resource are configured
"Hack" ansible steps of packer provision, to intercept the apt source to Artifactory.
because EMC Artifactory CA setup, it requires additional steps to add CA cetification.
```
jobs/build_docker/build_docker.groovy
jobs/build_docker/build_docker.sh
jobs/build_ova/build_ova.groovy
jobs/build_ova/build_ova.sh
jobs/build_vagrant/build_vagrant.groovy
jobs/build_vagrant/build_vagrant.sh
```

(5) the *deb cache* should also be changed.
image
code change in 
```
build-release-tools/application/make_debian_packages.py
```

**Test Done**
Experimental pipeline running on SH Jenkins sandbox  http://10.**.**.175:8080/job/Master-CI-RAC-4783/
it has been run for 8 days , about ~50 runs (good builds starts from build 66, the errors of later builds are due to different reason like sandbox vagrant issue . but the apt-tmp-repo functions are all good.)

**Reviewer**
@PengTian0  @changev   @johren

### CODE REVIEW

**Comment by panpan0000** [comment]
test this please

File: jobs/build_vagrant/build_vagrant.sh [hunk 1]
```diff
@@ -1,13 +1,28 @@
...
-ifconfig
 set +e
-packer -v
+ifconfig
+
+packer -v  # $? of "packer -v" is 1 ...
 vagrant -v
+set -e
+
+ARTIFACTORY_URL=http://afeossand1.cec.lab.emc.com/artifactory
```
Comment (PengTian0) [inline jobs/build_vagrant/build_vagrant.sh]: How about make it as a environment variable?

File: build-release-tools/application/make_debian_packages.py [hunk 2]
```diff
@@ -278,12 +335,18 @@ def main():
...
     bintray = Bintray(args.bintray_credential, args.bintray_subject, args.bintray_repo)
+    # Artifactory is the staging/temp place to store the deb between deb-build and post-test steps
+    artifactory = JFrogArtifactory( user_cred=(args.artifactory_username, args.artifactory_password), artifactory_loc= args.artifactory_url)
 
     checkout_repos(args.manifest_file, args.build_directory, args.force, args.jobs, git_credential=args.git_credential)
     all_repos = get_build_repos(args.build_directory)
-    package_not_exist = create_packages_filter(bintray, args.build_directory, args.is_official_release)
+    package_not_exist = create_packages_filter(bintray, artifactory, args.artifactory_repo, args.build_directory, args.is_official_release)
     package_need_build_repos = filter(package_not_exist, all_repos)
+    package_need_build_repos.append('RackHD') #always rebuild rackhd.deb
```
Comment (changev) [inline build-release-tools/application/make_debian_packages.py]: RackHD version is a daily timestamp it'll change everytime.
So package_not_exist won't contains rackhd.
If append "rackhd" again, the  package_need_build_repos will have 2 rackhd?

File: jobs/build_debian/build_debian.sh [hunk 3]
```diff
@@ -2,9 +2,12 @@
 set -ex
 pushd $WORKSPACE
 
+LOCAL_BUILD_DIR=b
+ARTIFACTORY_URL=http://afeossand1.cec.lab.emc.com/artifactory
```
Comment (changev) [inline jobs/build_debian/build_debian.sh]: Agree with @PengTian0 , make ARTIFACTORY_URL as an env var will be more convenient

**Comment by PengTian0** [comment]
@panpan0000 The job  packer cache build is using the temp bintray. Please update it to use the Internal Artifactory as the Debian Staging

File: build-release-tools/application/make_debian_packages.py (hunk 2 above)
Comment (panpan0000) [reply build-release-tools/application/make_debian_packages.py]: so careful review.@changev !
I will remove it.
I checked from the console log
RackHD did show up twice:
```
15:51:36 [Info] on-taskgraph version 2.5.0-20170513UTC-b573ca4 no cache version available on Bintray. Do deb build.
15:51:36 [Info] Repo on-core will run deb-build 
15:51:36 [Info] Repo on-http will run deb-build 
15:51:36 [Info] Repo on-statsd will run deb-build 
15:51:36 [Info] Repo on-tftp will run deb-build 
15:51:36 [Info] Repo di.js will run deb-build 
15:51:36 [Info] Repo on-wss will run deb-build 
15:51:36 [Info] Repo on-skupack will run deb-build 
15:51:36 [Info] Repo on-tasks will run deb-build 
15:51:36 [Info] Repo ucs-service will run deb-build 
15:51:36 [Info] Repo RackHD will run deb-build 
15:51:36 [Info] Repo on-syslog will run deb-build 
15:51:36 [Info] Repo on-dhcp-proxy will run deb-build 
15:51:36 [Info] Repo on-taskgraph will run deb-build 
15:51:36 [Info] Repo RackHD will run deb-build 
```

**Comment by PengTian0** [comment]
:+1:

**Comment by changev** [comment]
@panpan0000 
Please resolve the conflicts and newly run several rounds of tests to make sure this feature is still valid.
:+1: **After that.**

File: build-release-tools/application/make_debian_packages.py [hunk 4]
```diff
@@ -278,12 +335,17 @@ def main():
...
     args = parse_args(sys.argv[1:])
+    # Bintray is the offical/final place to store the deb
     bintray = Bintray(args.bintray_credential, args.bintray_subject, args.bintray_repo)
+    # Artifactory is the staging/temp place to store the deb between deb-build and post-test steps
+    artifactory = JFrogArtifactory( user_cred=(args.artifactory_username, args.artifactory_password), artifactory_loc= args.artifactory_url)
 
     checkout_repos(args.manifest_file, args.build_directory, args.force, args.jobs, git_credential=args.git_credential)
     all_repos = get_build_repos(args.build_directory)
-    package_not_exist = create_packages_filter(bintray, args.build_directory, args.is_official_release)
+    package_not_exist = create_packages_filter(bintray, artifactory, args.artifactory_repo, args.build_directory, args.is_official_release)
```
Comment (changev) [inline build-release-tools/application/make_debian_packages.py:345]: "rackhd" must in package_not_exist 
http://10.62.59.175:8080/blue/organizations/jenkins/Master-CI-RAC-4783/detail/Master-CI-RAC-4783/382/pipeline

**Comment by panpan0000** [comment]
Functionality has been proved in Sandbox Jenkins:

http://10.6x.5x.1xx:8080/job/Master-CI-RAC-4783/

Build:  497,530,535 are good

File: jobs/build_debian/build_debian.sh [hunk 5]
```diff
@@ -13,16 +16,22 @@ curl --user $BINTRAY_CREDS -L "$MANIFEST_FILE_URL" -o rackhd-manifest
 --is-official-release $IS_OFFICIAL_RELEASE \
 --bintray-credential BINTRAY_CREDS \
 --bintray-subject $BINTRAY_SUBJECT \
---bintray-repo $BINTRAY_REPO
+--bintray-repo $BINTRAY_REPO \
+--artifactory-url $ARTIFACTORY_URL \
+--artifactory-repo $STAGE_REPO_NAME \
+--artifactory-username $ARTIFACTORY_USR \
+--artifactory-password $ARTIFACTORY_PWD \
```
Comment (PengTian0) [inline jobs/build_debian/build_debian.sh:23]: The last line should remove ```\```

**Comment by panpan0000** [comment]
for MasterCI Job/SprintRelease Job:
CI_BINTRAY_SUBJECT can be removed.
but 3 new parameters should be added.

- ARTIFACTORY_URL
- STAGE_REPO_NAME
-DEB_COMPONENT

**Comment by PengTian0** [comment]
:+1:

**Comment by panpan0000** [comment]
http://rackhdci.lss.emc.com/job/Master-CI-2/
is the correct result runs

**Review by anhou — APPROVED** [review]

**Pull request merged**